source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -495,6 +495,7 @@ DnsResolverImpl::AddrInfoPendingResolution::availableInterfaces() {
 // c-ares DNS resolver factory
 class CaresDnsResolverFactory : public DnsResolverFactory {
 public:
+  CaresDnsResolverFactory() : ares_library_initialized_(false) {}
```

Comment:
are we guaranteed to have only one instance of CaresDnsResolverFactory ?

Reply:
Yes, IIUC, there is only one instance of CaresDnsResolverFactory(), which is registered here:
https://github.com/envoyproxy/envoy/blob/326b205af6b16dbf75f6bfbf246082f5966bc0b0/source/extensions/network/dns_resolver/cares/dns_impl.cc#L527

Every time when we need to createDnsResolver, we first retrieve the factory by, e.g:
https://github.com/envoyproxy/envoy/blob/326b205af6b16dbf75f6bfbf246082f5966bc0b0/source/server/server.cc#L678

Reply:
It's not obvious to me from looking at the code that there's only one instance of it.

I think it would be better to keep a separate structure that "owns" the initialization/shutdown of c_ares, which is created with MUTABLE_CONSTRUCT_ON_FIRST_USE, and provides start() and finish() methods that update a counter protected by a mutex, and have the init/destruct run inside the mutex on transitions of the count from 0->1 and 1->0.